xpu: support xpu backend from stock pytorch (>=2.4)

dvrogozh opened the pull request:
Fixes: https://github.com/huggingface/transformers/issues/31237

XPU backend is available in the stock PyTorch starting from version 2.4, see [1]. This commit extends huggingface transformers to support XPU from both IPEX and the stock pytorch. IPEX is being tried first.

Raising this PR as WIP and Draft to facilitate further discussion around XPU backend enabling in huggingface and be able to communicate observed XPU issues back to PyTorch.

See: https://github.com/pytorch/pytorch/issues/114842
Requires: https://github.com/huggingface/accelerate/pull/2825

cc: @muellerzr, @EikanWang, @jgong5, @kding1, @sywangyi

dvrogozh committed:
xpu: support xpu backend from stock pytorch (>=2.4)

Fixes: https://github.com/huggingface/transformers/issues/31237

XPU backend is available in the stock PyTorch starting from
version 2.4, see [1]. This commit extends huggingface transformers
to support XPU from both IPEX and the stock pytorch. IPEX is being
tried first.

See: https://github.com/pytorch/pytorch/issues/114842
Requires: https://github.com/huggingface/accelerate/pull/2825
Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

jgong5 commented on src/transformers/utils/import_utils.py:
```diff
@@ -747,13 +747,16 @@ def get_major_and_minor_from_version(full_version):
+        import intel_extension_for_pytorch  # noqa: F401
     import torch
 
+    #torch.set_default_device('xpu:0')
```
nit: remove it.

dvrogozh replied:
done

muellerzr reviewed (commented):
Great! This PR should be merged in-tandem with the accelerate one here: https://github.com/huggingface/accelerate/pull/2825

ydshieh commented:
Happy to see this addition 🚀 ! Just wondering if now we should be careful with the names like `Pytorch XPU` and `IPEX XPU`. (and sorry if this doesn't make sense 😅 )

ydshieh commented on src/transformers/utils/import_utils.py:
```diff
@@ -747,13 +747,18 @@ def get_major_and_minor_from_version(full_version):
-    import intel_extension_for_pytorch  # noqa: F401
     import torch
 
+    if is_ipex_available():
+        import intel_extension_for_pytorch  # noqa: F401
```
does this mean if IPEX is available, we will use XPU from there instead of from Pytorch? Is this what is preferred?

dvrogozh commented on src/transformers/testing_utils.py:
```diff
@@ -813,23 +813,24 @@ def require_torch_multi_npu(test_case):
 def require_torch_xpu(test_case):
     """
-    Decorator marking a test that requires XPU and IPEX.
+    Decorator marking a test that requires XPU (in PyTorch).
```
Let me try to clarify a context. There are 2 dimensions here:
1. Compute backend which is getting used, that's CPU or XPU (or cuda, mps, etc.)
2. Framework/library which exposes this backend

For the 2nd item (framework/library) here are 3 choices which are now available for Intel:
* IPEX CPU-only backend, i.e. IPEX flavor which exposes CPU optimizations
* IPEX XPU backend, i.e. IPEX which exposes support for Intel GPU which is called XPU backend
* Stock Pytorch >=2.4 which exposes XPU backend

So, what changed is that previously XPU backend implied IPEX-XPU installation, but now XPU backend might imply either IPEX-XPU installation or user just using Pytorch >=2.4.

Considering my change I just dropped framework/library implication which does not hold true anymore. I think the caller cares about backend availability disrespect to the installation mechanism which exposed this backend.

ydshieh replied:
OK. Thanks. But just wondering why not just `Decorator marking a test that requires XPU`, i.e. no `(in PyTorch)` part.

It's nit and I don't feel strong: I am just curious.

dvrogozh replied:
If you will open up this file, you'll notice few other similar places:
```
Decorator marking a test that requires NeuronCore (in PyTorch).
Decorator marking a test that requires NPU (in PyTorch).
...
```
So, I just aligned to other places actually. I guess that this `(in PyTorch)` is due to the point that this python code suites not only pytorch, but also tensorflow and maybe few other frameworks. As such this makes sense to emphasize area of method applicability.

ydshieh replied:
ah, ok. Thanks for explaining. Review next week!

dvrogozh commented:
I tried this PR (+ https://github.com/huggingface/accelerate/pull/2825 on which it depends) as much as I could in the IPEX-CPU, IPEX-XPU, Pytorch-XPU, Pytorch-CPU scenarios. Tried to run some tests from accelerate and transformers and some examples from transformers. All seem to work engaging with XPU when expected. I promote these PRs from drafts for the qualified review. Let me know if any concerns or any feedback needs to be addressed.

dvrogozh committed:
xpu: enable gpt2 and decision_transformer tests for xpu pytorch backend

Note that running xpu tests requires TRANSFORMERS_TEST_DEVICE_SPEC=spec.py
passed to the test runner:

  import torch
  DEVICE_NAME = 'xpu'
  MANUAL_SEED_FN = torch.xpu.manual_seed
  EMPTY_CACHE_FN = torch.xpu.empty_cache
  DEVICE_COUNT_FN = torch.xpu.device_count

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh commented:
I added one more commit to enable some tests for xpu backend.

dvrogozh commented:
Applied `python utils/check_copies.py --fix_and_overwrite` to propagate change in gpt2 to decision_transformer. This fixes failure noted by ci. Test for the later passes for xpu backend.

muellerzr reviewed (commented):
Thanks! Overall this looks fine to make, and makes sense why we need to adjust `models/decision_transformer/...` (to get the ipex patches in).

PR has been merged on the `accelerate` side, overall this seems good to me however: should we limit the `accelerate` version required for the xpu support to the new accelerate version? (to come out next month)

cc @amyeroberts

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31238). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

dvrogozh commented:
> however: should we limit the accelerate version required for the xpu support to the new accelerate version? (to come out next month)

@muellerzr : By bumping accelerate version to 0.32.0?
https://github.com/huggingface/transformers/blob/c624d5ba0b41542eb3daa0e2fdc985231551d776/setup.py#L99

muellerzr commented:
> By bumping accelerate version to 0.32.0?

We most certainly shouldn't do that :)

Actually, I think we're fairly okay, as accelerate will do a passthrough and IIUC this PR doesn't break old behavior, correct? (Basically, per my understanding if users run an old accelerate version nothing will break, right?)

dvrogozh commented:
> Actually, I think we're fairly okay, as accelerate will do a passthrough and IIUC this PR doesn't break old behavior, correct?

Yes, i think so. Till users are within previous usages (with IPEXes) nothing should change for them and be compatible.

> The question is if we should have a flag for a minimum accelerate version if they are on the xpu branch/logic

New accelerate is indeed required on xpu branch, otherwise there will be runtime error. So check will be useful. **I will add.**

dvrogozh commented on src/transformers/utils/import_utils.py:
```diff
@@ -747,13 +747,23 @@ def get_major_and_minor_from_version(full_version):
 
+    if is_ipex_available():
+        import intel_extension_for_pytorch  # noqa: F401
+    elif not is_accelerate_available("0.32.0.dev"):
```
@muellerzr : I added this check, since python version comparison evaluates `0.32.0.dev0 >= 0.32.0` as `False`, so I compared with `0.32.0.dev` and I am not sure that this is correct way. Please, advice.

dvrogozh replied:
and... I obviously can't raise exception here since all the tests will fail in this case. Problem is that returning `False` here does not give much clarity to the end user - all fails at runtime with pretty the same error as with having wrong accelerate.

amyeroberts commented on src/transformers/testing_utils.py:
```diff
@@ -2428,9 +2429,19 @@ def _device_agnostic_dispatch(device: str, dispatch_table: Dict[str, Callable],
+    BACKEND_MANUAL_SEED = {
+        "cpu": torch.manual_seed,
+        "cuda": torch.cuda.manual_seed,
+        "xpu": torch.xpu.manual_seed,
```
For what version of torch is `torch.xpu` available? We might need to guard this

dvrogozh replied:
`torch.xpu` is available from torch `2.3` (was initially added for IPEX).

Note: stock pytorch xpu support (w/o IPEX) is available from `2.4`.

amyeroberts replied:
OK, in this case, we can only add these values to the dictionaries if the available torch version >= 2.3.

dvrogozh replied:
Sure, I will make the change.

dvrogozh replied:
moved under `if is_torch_xpu_available()`

amyeroberts reviewed (commented):
Thanks for adding this! 

Just one question on the availability of `torch.xpu` which we might have to take care of in the testing utils

dvrogozh commented on src/transformers/utils/import_utils.py:
```diff
@@ -747,13 +747,20 @@ def get_major_and_minor_from_version(full_version):
 
+    if is_ipex_available():
+        import intel_extension_for_pytorch  # noqa: F401
+    elif not is_accelerate_available("0.33.0.dev"):
```
@muellerzr : I added this check, since python version comparison evaluates `0.32.0.dev0 >= 0.32.0` as False, so I compared with `0.32.0.dev` and I am not sure that this is correct way. Please, advice.

Also, I can't raise exception here since `is_torch_xpu_available` is on a generic path and will fail non-xpu cases. And without exception the error which end user gets looks quite similar to what he will get running with wrong accelerate version. Do you have an idea where to raise exception notifying user that accelerate version is wrong?

dvrogozh replied:
Ok, so, I remove accelerate check from `is_torch_xpu_available()` since I thought this function does its job and no need to add this here. And I added check with raising exception in here:
https://github.com/huggingface/transformers/blob/b7672826cad31e30319487af876e608d8af7d37b/src/transformers/training_args.py#L2138-L2140

amyeroberts reviewed (approved):
LGTM - thanks for enabling this!

muellerzr reviewed (approved):
Thanks! LG2M as well :) (We can revisit the 0.32.0.dev after 0.32.0 is out, I'll keep it in my notes)

faaany commented on src/transformers/testing_utils.py:
```diff
@@ -2431,6 +2432,10 @@ def _device_agnostic_dispatch(device: str, dispatch_table: Dict[str, Callable],
     BACKEND_EMPTY_CACHE = {"cuda": torch.cuda.empty_cache, "cpu": None, "default": None}
     BACKEND_DEVICE_COUNT = {"cuda": torch.cuda.device_count, "cpu": lambda: 0, "default": lambda: 1}
+    if is_torch_xpu_available():
+        BACKEND_MANUAL_SEED["xpu"] = torch.xpu.manual_seed
```
this should be removed, more info see https://huggingface.co/docs/transformers/v4.41.3/en/testing#testing-with-a-specific-pytorch-backend-or-device

dvrogozh replied:
@faaany : thank you for the link. I overlooked it. I removed these lines and used spec.py on my side.

ydshieh commented:
Hi @dvrogozh

Thank you for this support. As mentioned by @faaany and [my comment](https://github.com/huggingface/transformers/pull/31402#issuecomment-2168013905), it would be better not to include things like

> BACKEND_MANUAL_SEED["xpu"]
> BACKEND_EMPTY_CACHE["xpu"]

in this PR (so far) and let the user to have them define when they need to use other device.

Happy to revise that design in a separate PR.

dvrogozh commented:
> As mentioned by @faaany and https://github.com/huggingface/transformers/pull/31402#issuecomment-2168013905, it would be better not to include things like

@ydshieh :  removed. Used `TRANSFORMERS_TEST_DEVICE_SPEC=spec.py` on my side:
```
import torch

DEVICE_NAME = 'xpu'

MANUAL_SEED_FN = torch.xpu.manual_seed
EMPTY_CACHE_FN = torch.xpu.empty_cache
DEVICE_COUNT_FN = torch.xpu.device_count
```

ydshieh commented:
For 

> contains custom code which must be executed to correctly load the dataset

not related to this PR.

And you can rebase on main to include #31407 that will make that failure disappear

ydshieh reviewed (approved):
Thanks again. I think everything runs smooth on your side, right? Will merge once I get a confirmation from your side 💯 !

dvrogozh commented:
> And you can rebase on main to include https://github.com/huggingface/transformers/pull/31407 that will make that failure disappear

Hm. Code is already on top of latest master and includes #31407.

> I think everything runs smooth on your side, right? 

For xpu backend w/ spec.py? yes. And I ran non-xpu stuff as much as I locally could.

ydshieh commented:
OK, I will check CI. Thank you again for contributing 💯 !

dvrogozh commented:
Thank you for the merge. If all will be good, should land in `transformers==4.42.0` as far as I could tell.
